lib/c-api/src/wasm_c_api/wasi/mod.rs

```diff
@@ -304,7 +305,124 @@ pub unsafe extern "C" fn wasi_get_wasi_version(module: &wasm_module_t) -> wasi_v
         .unwrap_or(wasi_version_t::InvalidVersion)
 }
 
-/// Takes ownership of `wasi_env_t`.
+/// Non-standard type wrapping `wasm_extern_t` with the addition of
+/// two `wasm_name_t` respectively for the module name and the name of
+/// the extern (very likely to be an import). This non-standard type
+/// is used by the non-standard `wasi_get_unordered_imports` function.
+#[allow(non_camel_case_types)]
+#[derive(Clone)]
+pub struct wasm_named_extern_t {
+    module: Box<wasm_name_t>,
+    name: Box<wasm_name_t>,
+    r#extern: Box<wasm_extern_t>,
+}
+
+wasm_declare_boxed_vec!(named_extern);
+
+/// Non-standard function to get the module name of a
+/// `wasm_named_extern_t`.
+#[no_mangle]
+pub extern "C" fn wasm_named_extern_module(named_extern: &wasm_named_extern_t) -> &wasm_name_t {
+    named_extern.module.as_ref()
+}
+
+/// Non-standard function to get the name of a `wasm_named_extern_t`.
+#[no_mangle]
+pub extern "C" fn wasm_named_extern_name(named_extern: &wasm_named_extern_t) -> &wasm_name_t {
+    named_extern.name.as_ref()
+}
+
+/// Non-standard function to get the wrapped extern of a
+/// `wasm_named_extern_t`.
+#[no_mangle]
+pub extern "C" fn wasm_named_extern_extern(named_extern: &wasm_named_extern_t) -> &wasm_extern_t {
+    named_extern.r#extern.as_ref()
+}
+
+/// Non-standard function to get the imports needed for the WASI
+/// implementation with no particular order. Each import has its
+/// associated module name and name, so that it can be re-order later
+/// based on the `wasm_module_t` requirements.
+///
+/// This function takes ownership of `wasm_env_t`.
+#[no_mangle]
+pub unsafe extern "C" fn wasi_get_unordered_imports(
+    store: &wasm_store_t,
+    module: &wasm_module_t,
+    wasi_env: &wasi_env_t,
+    imports: &mut wasm_named_extern_vec_t,
+) -> bool {
+    wasi_get_unordered_imports_inner(store, module, wasi_env, imports).is_some()
+}
+
+unsafe fn wasi_get_unordered_imports_inner(
+    store: &wasm_store_t,
+    module: &wasm_module_t,
+    wasi_env: &wasi_env_t,
+    imports: &mut wasm_named_extern_vec_t,
+) -> Option<()> {
+    let store = &store.inner;
+
+    let version = c_try!(
+        get_wasi_version(&module.inner, false).ok_or_else(|| CApiError {
+            msg: "could not detect a WASI version on the given module".to_string(),
+        })
+    );
+
+    let import_object = generate_import_object_from_env(store, wasi_env.inner.clone(), version);
+
+    *imports = import_object
+        .into_iter()
+        .map(|((module, name), export)| {
+            let module = {
+                let mut module: Box<str> = module.into_boxed_str();
+                let module_ptr = module.as_mut_ptr();
+                let module_len = module.bytes().len();
+                let module_inner = wasm_name_t {
+                    size: module_len,
+                    data: module_ptr,
+                };
+
+                Box::leak(module);
```

Comment:
It'd be nice if we had a conversion function for this, I think we have something like this for `Vec<T>` or something, it'd be fine to have a `From` / `Into` because this code needs to exist in multiple places and anything but a standard way to do it means we'll get weird inconsistencies or bugs

Reply:
Also as part of this, lets check the usage of `wasm_name_t`, where does it get freed? Are the strings usually host-owned? Need to make sure everything works together and we're meeting all the expectations of the standard types we're using

Reply:
It's unclear in this case. I think it's best to give the entire ownership to the caller in this case, so `imports: wasm_named_extern_vec_t` is owned by the caller.

We can free the entire vector with `wasm_named_extern_vec_delete`. However, in real code, we often need to keep a copy of the inner `wasm_extern_t` value. Problem is that `wasm_extern_copy` doesn't exist for the moment, I might implement it as part of this PR too.

Reply:
Done.